h2o-core/src/main/java/water/MRTask.java

```diff
@@ -624,23 +625,27 @@ private void setupLocal0() {
         // Call all the various map() calls that apply
         if(_profile!=null)
           _profile._userstart = System.currentTimeMillis();
-        if( _fr.vecs().length == 1 ) map(bvs[0]);
-        if( _fr.vecs().length == 2 ) map(bvs[0], bvs[1]);
-        if( _fr.vecs().length == 3 ) map(bvs[0], bvs[1], bvs[2]);
-        if( true                  )  map(bvs );
-        if( _output_types != null && _output_types.length == 1 ) { // convenience versions for cases with single output.
-          if( appendableChunks == null ) throw H2O.fail(); // Silence IdeaJ warnings
-          if( _fr.vecs().length == 1 ) map(bvs[0], appendableChunks[0]);
-          if( _fr.vecs().length == 2 ) map(bvs[0], bvs[1],appendableChunks[0]);
-          //if( _fr.vecs().length == 3 ) map(bvs[0], bvs[1], bvs[2],appendableChunks[0]);
-          //if( true                  )  map(bvs,    appendableChunks[0]);
+        int fr_vecs_length = _fr.vecs().length;
+        int outputs_length = _output_types == null? 0 : _output_types.length;
+        if (outputs_length == 0) {
+          if (fr_vecs_length == 1) map(bvs[0]);
+          if (fr_vecs_length == 2) map(bvs[0], bvs[1]);
+          if (fr_vecs_length == 3) map(bvs[0], bvs[1], bvs[2]);
+          map(bvs);
         }
-        if( _output_types != null && _output_types.length == 2) { // convenience versions for cases with 2 outputs (e.g split).
-          if( appendableChunks == null ) throw H2O.fail(); // Silence IdeaJ warnings
-          if( _fr.vecs().length == 1 ) map(bvs[0], appendableChunks[0],appendableChunks[1]);
-          //if( _fr.vecs().length == 2 ) map(bvs[0], bvs[1],appendableChunks[0],appendableChunks[1]);
-          //if( _fr.vecs().length == 3 ) map(bvs[0], bvs[1], bvs[2],appendableChunks[0],appendableChunks[1]);
-          if( true                  )  map(bvs,    appendableChunks[0],appendableChunks[1]);
+        if (outputs_length == 1) {  // convenience versions for cases with single output.
+          assert appendableChunks != null;
+          if (fr_vecs_length == 1) map(bvs[0], appendableChunks[0]);
+          if (fr_vecs_length == 2) map(bvs[0], bvs[1], appendableChunks[0]);
+          // if (fr_vecs_length == 3) map(bvs[0], bvs[1], bvs[2], appendableChunks[0]);
+          map(bvs, appendableChunks[0]);
+        }
+        if (outputs_length == 2) {  // convenience versions for cases with 2 outputs (e.g split).
+          assert appendableChunks != null;
+          if (fr_vecs_length == 1) map(bvs[0], appendableChunks[0], appendableChunks[1]);
+          // if (fr_vecs_length == 2) map(bvs[0], bvs[1], appendableChunks[0], appendableChunks[1]);
+          // if (fr_vecs_length == 3) map(bvs[0], bvs[1], bvs[2], appendableChunks[0], appendableChunks[1]);
+          map(bvs, appendableChunks[0], appendableChunks[1]);
         }
         map(bvs,appendableChunks);
```

Comment:
As far as I understand, it will depend on which version of `doAll` you call on the task. If you request `doAll` without any outputs, then only the first `map()` is called; if you request some outputs... then you're right -- both get called...